llarp/profiling.cpp

```diff
@@ -241,37 +241,32 @@ namespace llarp
   bool
   Profiling::Save(const fs::path fpath)
   {
-    std::shared_lock lock{m_ProfilesMutex};
-    size_t sz = (m_Profiles.size() * (RouterProfile::MaxSize + 32 + 8)) + 8;
+    const size_t sz = (m_Profiles.size() * (RouterProfile::MaxSize + 32 + 8)) + 8;
 
     std::vector<byte_t> tmp(sz, 0);
     llarp_buffer_t buf(tmp);
-    auto res = BEncodeNoLock(&buf);
-    if (res)
+
     {
-      buf.sz = buf.cur - buf.base;
-      auto optional_f = util::OpenFileStream<std::ofstream>(fpath, std::ios::binary);
-      if (!optional_f)
+      util::Lock lock{m_ProfilesMutex};
+      if (not BEncode(&buf))
         return false;
-      auto& f = *optional_f;
-      if (f.is_open())
-      {
-        f.write((char*)buf.base, buf.sz);
-        m_LastSave = llarp::time_now_ms();
-      }
     }
-    return res;
-  }
 
-  bool
-  Profiling::BEncode(llarp_buffer_t* buf) const
-  {
-    std::shared_lock lock{m_ProfilesMutex};
-    return BEncodeNoLock(buf);
+    buf.sz = buf.cur - buf.base;
+    auto optional_f = util::OpenFileStream<std::ofstream>(fpath, std::ios::binary);
+    if (!optional_f)
+      return false;
+    auto& f = *optional_f;
+    if (not f.is_open())
+      return false;
+
+    f.write(reinterpret_cast<const char*>(buf.base), buf.sz);
```

Comment:
I would assume this write can fail, right?  If so, should probably check for that since `m_LastSave` time is being set.